{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b118b01b-ac4d-4663-8280-8909461dc316",
   "metadata": {},
   "source": [
    "# Data Science and Machine Learning Individual Project"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c92ed29e-cf14-46ca-9e22-c63c4a262a33",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "The scripts in this notebook explore a dataset of bulk RNA seq data, following cells across a 5 day span under different cell culture conditions. The aim of the work that generated this data was to study the transition of human naive epiblast cells to trophectoderm, a transition that is not observed in the widely used mouse embryo models.\n",
    "\n",
    "\n",
    "## Cell Culture Conditions\n",
    "The standard cell culture medium for human naive pluripotent stem cells is PXGL, a culture medium made from the base N2B27 supplemented with PD0325901, XAV939, Go6983 and hLIF. These supplements help to block the differentiation of naive cells, allowing them to be maintained in culture. \n",
    "\n",
    "\n",
    "The bulk RNA seq data set was produced from whole-transcriptome sequencing over a 5 day time course in one of the following culture medium conditions:\n",
    "\n",
    "N2B27 (unsupplemented)\n",
    "\n",
    "PD0325901, hereafter PD03 (a MEK/ERK inhibitor) \n",
    "\n",
    "A83-01, (an Activin/Nodal inhibitor)\n",
    "\n",
    "PD03 and A83-01 combined.\n",
    "\n",
    "The paper in which this data was published showed that PD03 and A83-01 combined allow the differentiation of human naive epiblast cells to trophectoderm cells (Human naive epiblast cells possess unrestricted lineage potential, Guo et al (2021)). The data is provided in FPKM (fragments per kilobase of transcript per million fragments). "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6a76684-1e5f-4d11-acbe-41ed42f382fc",
   "metadata": {},
   "source": [
    "# Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9faa152-d763-47b0-9f95-470d3e415ffd",
   "metadata": {},
   "source": [
    "## Import relevant libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b524a24-e7be-422c-b692-8dfccaf404ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import numpy as np\n",
    "import re\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a9eb943-3ff6-4669-bb02-b8c344af3a4b",
   "metadata": {},
   "source": [
    "## Read in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "98044062-e61f-4cac-85cc-00582b319fcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>biotype</th>\n",
       "      <th>geneID</th>\n",
       "      <th>PDLGX_hNES1</th>\n",
       "      <th>A8301.D1_hNES1</th>\n",
       "      <th>A8301.D2_hNES1</th>\n",
       "      <th>A8301.D3_hNES1</th>\n",
       "      <th>N2B27.D1_hNES1</th>\n",
       "      <th>N2B27.D2_hNES1</th>\n",
       "      <th>N2B27.D3_hNES1</th>\n",
       "      <th>N2B27.D5_hNES1</th>\n",
       "      <th>PD.D1_hNES1</th>\n",
       "      <th>PD.D2_hNES1</th>\n",
       "      <th>PD.D3_hNES1</th>\n",
       "      <th>PD.D5_hNES1</th>\n",
       "      <th>PDA8301.D1_hNES1</th>\n",
       "      <th>PDA8301.D2_hNES1</th>\n",
       "      <th>PDA8301.D3_hNES1</th>\n",
       "      <th>PDA8301.D5_hNES1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>5.05</td>\n",
       "      <td>8.00</td>\n",
       "      <td>8.62</td>\n",
       "      <td>6.76</td>\n",
       "      <td>8.20</td>\n",
       "      <td>9.42</td>\n",
       "      <td>7.51</td>\n",
       "      <td>6.79</td>\n",
       "      <td>5.36</td>\n",
       "      <td>6.30</td>\n",
       "      <td>6.97</td>\n",
       "      <td>5.86</td>\n",
       "      <td>6.00</td>\n",
       "      <td>8.35</td>\n",
       "      <td>7.48</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>47.67</td>\n",
       "      <td>38.80</td>\n",
       "      <td>48.26</td>\n",
       "      <td>47.27</td>\n",
       "      <td>40.02</td>\n",
       "      <td>40.67</td>\n",
       "      <td>42.02</td>\n",
       "      <td>38.26</td>\n",
       "      <td>41.14</td>\n",
       "      <td>40.60</td>\n",
       "      <td>37.82</td>\n",
       "      <td>38.43</td>\n",
       "      <td>43.27</td>\n",
       "      <td>39.45</td>\n",
       "      <td>37.40</td>\n",
       "      <td>34.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1.13</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.09</td>\n",
       "      <td>1.16</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.98</td>\n",
       "      <td>1.27</td>\n",
       "      <td>1.01</td>\n",
       "      <td>1.15</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.96</td>\n",
       "      <td>1.11</td>\n",
       "      <td>1.19</td>\n",
       "      <td>1.82</td>\n",
       "      <td>1.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1.87</td>\n",
       "      <td>1.98</td>\n",
       "      <td>2.32</td>\n",
       "      <td>2.40</td>\n",
       "      <td>2.12</td>\n",
       "      <td>2.04</td>\n",
       "      <td>2.52</td>\n",
       "      <td>2.55</td>\n",
       "      <td>2.10</td>\n",
       "      <td>2.31</td>\n",
       "      <td>3.36</td>\n",
       "      <td>3.73</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.54</td>\n",
       "      <td>3.93</td>\n",
       "      <td>3.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58879</th>\n",
       "      <td>ENSG00000286268</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AF196969.1</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58880</th>\n",
       "      <td>ENSG00000286269</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>AC092661.1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58881</th>\n",
       "      <td>ENSG00000286270</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>AL591074.2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58882</th>\n",
       "      <td>ENSG00000286271</td>\n",
       "      <td>antisense</td>\n",
       "      <td>AC008945.2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58883</th>\n",
       "      <td>ENSG00000286272</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>AC093843.2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58884 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ensemblID         biotype      geneID  PDLGX_hNES1  \\\n",
       "0      ENSG00000000003  protein_coding      TSPAN6         5.05   \n",
       "1      ENSG00000000005  protein_coding        TNMD         0.02   \n",
       "2      ENSG00000000419  protein_coding        DPM1        47.67   \n",
       "3      ENSG00000000457  protein_coding       SCYL3         1.13   \n",
       "4      ENSG00000000460  protein_coding    C1orf112         1.87   \n",
       "...                ...             ...         ...          ...   \n",
       "58879  ENSG00000286268  protein_coding  AF196969.1         0.32   \n",
       "58880  ENSG00000286269         lincRNA  AC092661.1         0.00   \n",
       "58881  ENSG00000286270         lincRNA  AL591074.2         0.00   \n",
       "58882  ENSG00000286271       antisense  AC008945.2         0.00   \n",
       "58883  ENSG00000286272         lincRNA  AC093843.2         0.00   \n",
       "\n",
       "       A8301.D1_hNES1  A8301.D2_hNES1  A8301.D3_hNES1  N2B27.D1_hNES1  \\\n",
       "0                8.00            8.62            6.76            8.20   \n",
       "1                0.08            0.07            0.03            0.12   \n",
       "2               38.80           48.26           47.27           40.02   \n",
       "3                1.10            1.10            1.09            1.16   \n",
       "4                1.98            2.32            2.40            2.12   \n",
       "...               ...             ...             ...             ...   \n",
       "58879            0.34            0.22            0.21            0.56   \n",
       "58880            0.00            0.00            0.00            0.00   \n",
       "58881            0.00            0.00            0.00            0.00   \n",
       "58882            0.01            0.00            0.01            0.03   \n",
       "58883            0.00            0.00            0.00            0.00   \n",
       "\n",
       "       N2B27.D2_hNES1  N2B27.D3_hNES1  N2B27.D5_hNES1  PD.D1_hNES1  \\\n",
       "0                9.42            7.51            6.79         5.36   \n",
       "1                0.13            0.12            0.04         0.14   \n",
       "2               40.67           42.02           38.26        41.14   \n",
       "3                1.05            0.98            1.27         1.01   \n",
       "4                2.04            2.52            2.55         2.10   \n",
       "...               ...             ...             ...          ...   \n",
       "58879            0.21            0.31            0.30         0.32   \n",
       "58880            0.01            0.00            0.00         0.00   \n",
       "58881            0.00            0.01            0.01         0.00   \n",
       "58882            0.00            0.01            0.06         0.00   \n",
       "58883            0.00            0.00            0.00         0.00   \n",
       "\n",
       "       PD.D2_hNES1  PD.D3_hNES1  PD.D5_hNES1  PDA8301.D1_hNES1  \\\n",
       "0             6.30         6.97         5.86              6.00   \n",
       "1             0.05         0.07         0.18              0.02   \n",
       "2            40.60        37.82        38.43             43.27   \n",
       "3             1.15         1.56         1.96              1.11   \n",
       "4             2.31         3.36         3.73              2.17   \n",
       "...            ...          ...          ...               ...   \n",
       "58879         0.27         0.35         0.47              0.26   \n",
       "58880         0.00         0.01         0.01              0.01   \n",
       "58881         0.00         0.00         0.00              0.00   \n",
       "58882         0.01         0.01         0.00              0.03   \n",
       "58883         0.00         0.00         0.00              0.00   \n",
       "\n",
       "       PDA8301.D2_hNES1  PDA8301.D3_hNES1  PDA8301.D5_hNES1  \n",
       "0                  8.35              7.48              6.67  \n",
       "1                  0.04              0.12              0.08  \n",
       "2                 39.45             37.40             34.68  \n",
       "3                  1.19              1.82              1.86  \n",
       "4                  2.54              3.93              3.13  \n",
       "...                 ...               ...               ...  \n",
       "58879              0.28              0.53              0.46  \n",
       "58880              0.00              0.00              0.00  \n",
       "58881              0.00              0.00              0.00  \n",
       "58882              0.00              0.01              0.00  \n",
       "58883              0.00              0.00              0.00  \n",
       "\n",
       "[58884 rows x 19 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = 'https://raw.githubusercontent.com/ehsd737/Data-Science-Assessment/main/BulkRNA_Seq_Data.csv'\n",
    "data = pd.read_csv(url)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f2680f3-6b7f-4d27-943b-6e8e4132b0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#data = pd.read_csv('C:/Users/ed737/OneDrive - University of Exeter/Documents/BulkRNA_Seq_Data.csv') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2a6e67a2-b021-4cc8-b5e7-1eba4cbb0fb9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>biotype</th>\n",
       "      <th>geneID</th>\n",
       "      <th>PDLGX_hNES1</th>\n",
       "      <th>A8301.D1_hNES1</th>\n",
       "      <th>A8301.D2_hNES1</th>\n",
       "      <th>A8301.D3_hNES1</th>\n",
       "      <th>N2B27.D1_hNES1</th>\n",
       "      <th>N2B27.D2_hNES1</th>\n",
       "      <th>N2B27.D3_hNES1</th>\n",
       "      <th>N2B27.D5_hNES1</th>\n",
       "      <th>PD.D1_hNES1</th>\n",
       "      <th>PD.D2_hNES1</th>\n",
       "      <th>PD.D3_hNES1</th>\n",
       "      <th>PD.D5_hNES1</th>\n",
       "      <th>PDA8301.D1_hNES1</th>\n",
       "      <th>PDA8301.D2_hNES1</th>\n",
       "      <th>PDA8301.D3_hNES1</th>\n",
       "      <th>PDA8301.D5_hNES1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>5.05</td>\n",
       "      <td>8.00</td>\n",
       "      <td>8.62</td>\n",
       "      <td>6.76</td>\n",
       "      <td>8.20</td>\n",
       "      <td>9.42</td>\n",
       "      <td>7.51</td>\n",
       "      <td>6.79</td>\n",
       "      <td>5.36</td>\n",
       "      <td>6.30</td>\n",
       "      <td>6.97</td>\n",
       "      <td>5.86</td>\n",
       "      <td>6.00</td>\n",
       "      <td>8.35</td>\n",
       "      <td>7.48</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>47.67</td>\n",
       "      <td>38.80</td>\n",
       "      <td>48.26</td>\n",
       "      <td>47.27</td>\n",
       "      <td>40.02</td>\n",
       "      <td>40.67</td>\n",
       "      <td>42.02</td>\n",
       "      <td>38.26</td>\n",
       "      <td>41.14</td>\n",
       "      <td>40.60</td>\n",
       "      <td>37.82</td>\n",
       "      <td>38.43</td>\n",
       "      <td>43.27</td>\n",
       "      <td>39.45</td>\n",
       "      <td>37.40</td>\n",
       "      <td>34.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1.13</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.09</td>\n",
       "      <td>1.16</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.98</td>\n",
       "      <td>1.27</td>\n",
       "      <td>1.01</td>\n",
       "      <td>1.15</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.96</td>\n",
       "      <td>1.11</td>\n",
       "      <td>1.19</td>\n",
       "      <td>1.82</td>\n",
       "      <td>1.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1.87</td>\n",
       "      <td>1.98</td>\n",
       "      <td>2.32</td>\n",
       "      <td>2.40</td>\n",
       "      <td>2.12</td>\n",
       "      <td>2.04</td>\n",
       "      <td>2.52</td>\n",
       "      <td>2.55</td>\n",
       "      <td>2.10</td>\n",
       "      <td>2.31</td>\n",
       "      <td>3.36</td>\n",
       "      <td>3.73</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.54</td>\n",
       "      <td>3.93</td>\n",
       "      <td>3.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58879</th>\n",
       "      <td>ENSG00000286268</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AF196969.1</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58880</th>\n",
       "      <td>ENSG00000286269</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>AC092661.1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58881</th>\n",
       "      <td>ENSG00000286270</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>AL591074.2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58882</th>\n",
       "      <td>ENSG00000286271</td>\n",
       "      <td>antisense</td>\n",
       "      <td>AC008945.2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58883</th>\n",
       "      <td>ENSG00000286272</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>AC093843.2</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58884 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ensemblID         biotype      geneID  PDLGX_hNES1  \\\n",
       "0      ENSG00000000003  protein_coding      TSPAN6         5.05   \n",
       "1      ENSG00000000005  protein_coding        TNMD         0.02   \n",
       "2      ENSG00000000419  protein_coding        DPM1        47.67   \n",
       "3      ENSG00000000457  protein_coding       SCYL3         1.13   \n",
       "4      ENSG00000000460  protein_coding    C1orf112         1.87   \n",
       "...                ...             ...         ...          ...   \n",
       "58879  ENSG00000286268  protein_coding  AF196969.1         0.32   \n",
       "58880  ENSG00000286269         lincRNA  AC092661.1         0.00   \n",
       "58881  ENSG00000286270         lincRNA  AL591074.2         0.00   \n",
       "58882  ENSG00000286271       antisense  AC008945.2         0.00   \n",
       "58883  ENSG00000286272         lincRNA  AC093843.2         0.00   \n",
       "\n",
       "       A8301.D1_hNES1  A8301.D2_hNES1  A8301.D3_hNES1  N2B27.D1_hNES1  \\\n",
       "0                8.00            8.62            6.76            8.20   \n",
       "1                0.08            0.07            0.03            0.12   \n",
       "2               38.80           48.26           47.27           40.02   \n",
       "3                1.10            1.10            1.09            1.16   \n",
       "4                1.98            2.32            2.40            2.12   \n",
       "...               ...             ...             ...             ...   \n",
       "58879            0.34            0.22            0.21            0.56   \n",
       "58880            0.00            0.00            0.00            0.00   \n",
       "58881            0.00            0.00            0.00            0.00   \n",
       "58882            0.01            0.00            0.01            0.03   \n",
       "58883            0.00            0.00            0.00            0.00   \n",
       "\n",
       "       N2B27.D2_hNES1  N2B27.D3_hNES1  N2B27.D5_hNES1  PD.D1_hNES1  \\\n",
       "0                9.42            7.51            6.79         5.36   \n",
       "1                0.13            0.12            0.04         0.14   \n",
       "2               40.67           42.02           38.26        41.14   \n",
       "3                1.05            0.98            1.27         1.01   \n",
       "4                2.04            2.52            2.55         2.10   \n",
       "...               ...             ...             ...          ...   \n",
       "58879            0.21            0.31            0.30         0.32   \n",
       "58880            0.01            0.00            0.00         0.00   \n",
       "58881            0.00            0.01            0.01         0.00   \n",
       "58882            0.00            0.01            0.06         0.00   \n",
       "58883            0.00            0.00            0.00         0.00   \n",
       "\n",
       "       PD.D2_hNES1  PD.D3_hNES1  PD.D5_hNES1  PDA8301.D1_hNES1  \\\n",
       "0             6.30         6.97         5.86              6.00   \n",
       "1             0.05         0.07         0.18              0.02   \n",
       "2            40.60        37.82        38.43             43.27   \n",
       "3             1.15         1.56         1.96              1.11   \n",
       "4             2.31         3.36         3.73              2.17   \n",
       "...            ...          ...          ...               ...   \n",
       "58879         0.27         0.35         0.47              0.26   \n",
       "58880         0.00         0.01         0.01              0.01   \n",
       "58881         0.00         0.00         0.00              0.00   \n",
       "58882         0.01         0.01         0.00              0.03   \n",
       "58883         0.00         0.00         0.00              0.00   \n",
       "\n",
       "       PDA8301.D2_hNES1  PDA8301.D3_hNES1  PDA8301.D5_hNES1  \n",
       "0                  8.35              7.48              6.67  \n",
       "1                  0.04              0.12              0.08  \n",
       "2                 39.45             37.40             34.68  \n",
       "3                  1.19              1.82              1.86  \n",
       "4                  2.54              3.93              3.13  \n",
       "...                 ...               ...               ...  \n",
       "58879              0.28              0.53              0.46  \n",
       "58880              0.00              0.00              0.00  \n",
       "58881              0.00              0.00              0.00  \n",
       "58882              0.00              0.01              0.00  \n",
       "58883              0.00              0.00              0.00  \n",
       "\n",
       "[58884 rows x 19 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# some initial data exploration\n",
    "\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2a5ba9a1-4d00-47ec-8632-26c80a44330d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TSPAN6' 'TNMD' 'DPM1' ... 'AL591074.2' 'AC008945.2' 'AC093843.2']\n"
     ]
    }
   ],
   "source": [
    "gene_list = data[\"geneID\"].unique()\n",
    "print(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50aae2a4-1d32-4ef5-8dcf-38f9b1a782a0",
   "metadata": {},
   "source": [
    "# Prepare the dataframe for further analysis \n",
    "This involves removing all non-protein-coding gene data, and all zero-read data, and then log2-transforming the dataset. \n",
    "\n",
    "Zero read data must be removed as if not it will become undefined after applying the log2 transformation. Alternatively,\n",
    "a small constant could be added prior to log transformation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "52d34980-6d7e-44fb-9efd-4ce7a6f73d1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>biotype</th>\n",
       "      <th>geneID</th>\n",
       "      <th>PDLGX_hNES1</th>\n",
       "      <th>A8301.D1_hNES1</th>\n",
       "      <th>A8301.D2_hNES1</th>\n",
       "      <th>A8301.D3_hNES1</th>\n",
       "      <th>N2B27.D1_hNES1</th>\n",
       "      <th>N2B27.D2_hNES1</th>\n",
       "      <th>N2B27.D3_hNES1</th>\n",
       "      <th>N2B27.D5_hNES1</th>\n",
       "      <th>PD.D1_hNES1</th>\n",
       "      <th>PD.D2_hNES1</th>\n",
       "      <th>PD.D3_hNES1</th>\n",
       "      <th>PD.D5_hNES1</th>\n",
       "      <th>PDA8301.D1_hNES1</th>\n",
       "      <th>PDA8301.D2_hNES1</th>\n",
       "      <th>PDA8301.D3_hNES1</th>\n",
       "      <th>PDA8301.D5_hNES1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>5.05</td>\n",
       "      <td>8.00</td>\n",
       "      <td>8.62</td>\n",
       "      <td>6.76</td>\n",
       "      <td>8.20</td>\n",
       "      <td>9.42</td>\n",
       "      <td>7.51</td>\n",
       "      <td>6.79</td>\n",
       "      <td>5.36</td>\n",
       "      <td>6.30</td>\n",
       "      <td>6.97</td>\n",
       "      <td>5.86</td>\n",
       "      <td>6.00</td>\n",
       "      <td>8.35</td>\n",
       "      <td>7.48</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>47.67</td>\n",
       "      <td>38.80</td>\n",
       "      <td>48.26</td>\n",
       "      <td>47.27</td>\n",
       "      <td>40.02</td>\n",
       "      <td>40.67</td>\n",
       "      <td>42.02</td>\n",
       "      <td>38.26</td>\n",
       "      <td>41.14</td>\n",
       "      <td>40.60</td>\n",
       "      <td>37.82</td>\n",
       "      <td>38.43</td>\n",
       "      <td>43.27</td>\n",
       "      <td>39.45</td>\n",
       "      <td>37.40</td>\n",
       "      <td>34.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1.13</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.09</td>\n",
       "      <td>1.16</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.98</td>\n",
       "      <td>1.27</td>\n",
       "      <td>1.01</td>\n",
       "      <td>1.15</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.96</td>\n",
       "      <td>1.11</td>\n",
       "      <td>1.19</td>\n",
       "      <td>1.82</td>\n",
       "      <td>1.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1.87</td>\n",
       "      <td>1.98</td>\n",
       "      <td>2.32</td>\n",
       "      <td>2.40</td>\n",
       "      <td>2.12</td>\n",
       "      <td>2.04</td>\n",
       "      <td>2.52</td>\n",
       "      <td>2.55</td>\n",
       "      <td>2.10</td>\n",
       "      <td>2.31</td>\n",
       "      <td>3.36</td>\n",
       "      <td>3.73</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.54</td>\n",
       "      <td>3.93</td>\n",
       "      <td>3.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58850</th>\n",
       "      <td>ENSG00000286239</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC093884.1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58872</th>\n",
       "      <td>ENSG00000286261</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC022137.3</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1.18</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.14</td>\n",
       "      <td>1.12</td>\n",
       "      <td>1.20</td>\n",
       "      <td>1.19</td>\n",
       "      <td>1.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58875</th>\n",
       "      <td>ENSG00000286264</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AP001453.5</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58876</th>\n",
       "      <td>ENSG00000286265</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC007244.1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58879</th>\n",
       "      <td>ENSG00000286268</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AF196969.1</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19997 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ensemblID         biotype      geneID  PDLGX_hNES1  \\\n",
       "0      ENSG00000000003  protein_coding      TSPAN6         5.05   \n",
       "1      ENSG00000000005  protein_coding        TNMD         0.02   \n",
       "2      ENSG00000000419  protein_coding        DPM1        47.67   \n",
       "3      ENSG00000000457  protein_coding       SCYL3         1.13   \n",
       "4      ENSG00000000460  protein_coding    C1orf112         1.87   \n",
       "...                ...             ...         ...          ...   \n",
       "58850  ENSG00000286239  protein_coding  AC093884.1         0.00   \n",
       "58872  ENSG00000286261  protein_coding  AC022137.3         0.86   \n",
       "58875  ENSG00000286264  protein_coding  AP001453.5         0.03   \n",
       "58876  ENSG00000286265  protein_coding  AC007244.1         0.00   \n",
       "58879  ENSG00000286268  protein_coding  AF196969.1         0.32   \n",
       "\n",
       "       A8301.D1_hNES1  A8301.D2_hNES1  A8301.D3_hNES1  N2B27.D1_hNES1  \\\n",
       "0                8.00            8.62            6.76            8.20   \n",
       "1                0.08            0.07            0.03            0.12   \n",
       "2               38.80           48.26           47.27           40.02   \n",
       "3                1.10            1.10            1.09            1.16   \n",
       "4                1.98            2.32            2.40            2.12   \n",
       "...               ...             ...             ...             ...   \n",
       "58850            0.00            0.00            0.00            0.00   \n",
       "58872            0.90            0.33            0.58            0.73   \n",
       "58875            0.02            0.02            0.01            0.03   \n",
       "58876            0.11            0.03            0.07            0.09   \n",
       "58879            0.34            0.22            0.21            0.56   \n",
       "\n",
       "       N2B27.D2_hNES1  N2B27.D3_hNES1  N2B27.D5_hNES1  PD.D1_hNES1  \\\n",
       "0                9.42            7.51            6.79         5.36   \n",
       "1                0.13            0.12            0.04         0.14   \n",
       "2               40.67           42.02           38.26        41.14   \n",
       "3                1.05            0.98            1.27         1.01   \n",
       "4                2.04            2.52            2.55         2.10   \n",
       "...               ...             ...             ...          ...   \n",
       "58850            0.00            0.00            0.00         0.00   \n",
       "58872            0.59            0.76            0.74         1.18   \n",
       "58875            0.03            0.00            0.01         0.02   \n",
       "58876            0.00            0.07            0.00         0.04   \n",
       "58879            0.21            0.31            0.30         0.32   \n",
       "\n",
       "       PD.D2_hNES1  PD.D3_hNES1  PD.D5_hNES1  PDA8301.D1_hNES1  \\\n",
       "0             6.30         6.97         5.86              6.00   \n",
       "1             0.05         0.07         0.18              0.02   \n",
       "2            40.60        37.82        38.43             43.27   \n",
       "3             1.15         1.56         1.96              1.11   \n",
       "4             2.31         3.36         3.73              2.17   \n",
       "...            ...          ...          ...               ...   \n",
       "58850         0.00         0.01         0.00              0.00   \n",
       "58872         0.89         1.10         1.14              1.12   \n",
       "58875         0.03         0.04         0.01              0.01   \n",
       "58876         0.00         0.00         0.11              0.21   \n",
       "58879         0.27         0.35         0.47              0.26   \n",
       "\n",
       "       PDA8301.D2_hNES1  PDA8301.D3_hNES1  PDA8301.D5_hNES1  \n",
       "0                  8.35              7.48              6.67  \n",
       "1                  0.04              0.12              0.08  \n",
       "2                 39.45             37.40             34.68  \n",
       "3                  1.19              1.82              1.86  \n",
       "4                  2.54              3.93              3.13  \n",
       "...                 ...               ...               ...  \n",
       "58850              0.01              0.00              0.00  \n",
       "58872              1.20              1.19              1.19  \n",
       "58875              0.00              0.01              0.00  \n",
       "58876              0.03              0.11              0.43  \n",
       "58879              0.28              0.53              0.46  \n",
       "\n",
       "[19997 rows x 19 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove all non protein-coding data\n",
    "# while this could be interesting to analyse, this is not the core focus of this analysis\n",
    "\n",
    "filtered_data = data[data[\"biotype\"].str.contains(\"protein_coding\", na=False)]\n",
    "\n",
    "filtered_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e317998f-110c-4580-9623-81787ea0cc74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>biotype</th>\n",
       "      <th>geneID</th>\n",
       "      <th>PDLGX_hNES1</th>\n",
       "      <th>A8301.D1_hNES1</th>\n",
       "      <th>A8301.D2_hNES1</th>\n",
       "      <th>A8301.D3_hNES1</th>\n",
       "      <th>N2B27.D1_hNES1</th>\n",
       "      <th>N2B27.D2_hNES1</th>\n",
       "      <th>N2B27.D3_hNES1</th>\n",
       "      <th>N2B27.D5_hNES1</th>\n",
       "      <th>PD.D1_hNES1</th>\n",
       "      <th>PD.D2_hNES1</th>\n",
       "      <th>PD.D3_hNES1</th>\n",
       "      <th>PD.D5_hNES1</th>\n",
       "      <th>PDA8301.D1_hNES1</th>\n",
       "      <th>PDA8301.D2_hNES1</th>\n",
       "      <th>PDA8301.D3_hNES1</th>\n",
       "      <th>PDA8301.D5_hNES1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>5.05</td>\n",
       "      <td>8.00</td>\n",
       "      <td>8.62</td>\n",
       "      <td>6.76</td>\n",
       "      <td>8.20</td>\n",
       "      <td>9.42</td>\n",
       "      <td>7.51</td>\n",
       "      <td>6.79</td>\n",
       "      <td>5.36</td>\n",
       "      <td>6.30</td>\n",
       "      <td>6.97</td>\n",
       "      <td>5.86</td>\n",
       "      <td>6.00</td>\n",
       "      <td>8.35</td>\n",
       "      <td>7.48</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>47.67</td>\n",
       "      <td>38.80</td>\n",
       "      <td>48.26</td>\n",
       "      <td>47.27</td>\n",
       "      <td>40.02</td>\n",
       "      <td>40.67</td>\n",
       "      <td>42.02</td>\n",
       "      <td>38.26</td>\n",
       "      <td>41.14</td>\n",
       "      <td>40.60</td>\n",
       "      <td>37.82</td>\n",
       "      <td>38.43</td>\n",
       "      <td>43.27</td>\n",
       "      <td>39.45</td>\n",
       "      <td>37.40</td>\n",
       "      <td>34.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1.13</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.09</td>\n",
       "      <td>1.16</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.98</td>\n",
       "      <td>1.27</td>\n",
       "      <td>1.01</td>\n",
       "      <td>1.15</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.96</td>\n",
       "      <td>1.11</td>\n",
       "      <td>1.19</td>\n",
       "      <td>1.82</td>\n",
       "      <td>1.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1.87</td>\n",
       "      <td>1.98</td>\n",
       "      <td>2.32</td>\n",
       "      <td>2.40</td>\n",
       "      <td>2.12</td>\n",
       "      <td>2.04</td>\n",
       "      <td>2.52</td>\n",
       "      <td>2.55</td>\n",
       "      <td>2.10</td>\n",
       "      <td>2.31</td>\n",
       "      <td>3.36</td>\n",
       "      <td>3.73</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.54</td>\n",
       "      <td>3.93</td>\n",
       "      <td>3.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58802</th>\n",
       "      <td>ENSG00000286190</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC055839.2</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58804</th>\n",
       "      <td>ENSG00000286192</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC069288.1</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58830</th>\n",
       "      <td>ENSG00000286219</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC242843.1</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.52</td>\n",
       "      <td>1.27</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.64</td>\n",
       "      <td>1.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58872</th>\n",
       "      <td>ENSG00000286261</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC022137.3</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1.18</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.14</td>\n",
       "      <td>1.12</td>\n",
       "      <td>1.20</td>\n",
       "      <td>1.19</td>\n",
       "      <td>1.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58879</th>\n",
       "      <td>ENSG00000286268</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AF196969.1</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15454 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ensemblID         biotype      geneID  PDLGX_hNES1  \\\n",
       "0      ENSG00000000003  protein_coding      TSPAN6         5.05   \n",
       "1      ENSG00000000005  protein_coding        TNMD         0.02   \n",
       "2      ENSG00000000419  protein_coding        DPM1        47.67   \n",
       "3      ENSG00000000457  protein_coding       SCYL3         1.13   \n",
       "4      ENSG00000000460  protein_coding    C1orf112         1.87   \n",
       "...                ...             ...         ...          ...   \n",
       "58802  ENSG00000286190  protein_coding  AC055839.2         0.11   \n",
       "58804  ENSG00000286192  protein_coding  AC069288.1         0.02   \n",
       "58830  ENSG00000286219  protein_coding  AC242843.1         0.77   \n",
       "58872  ENSG00000286261  protein_coding  AC022137.3         0.86   \n",
       "58879  ENSG00000286268  protein_coding  AF196969.1         0.32   \n",
       "\n",
       "       A8301.D1_hNES1  A8301.D2_hNES1  A8301.D3_hNES1  N2B27.D1_hNES1  \\\n",
       "0                8.00            8.62            6.76            8.20   \n",
       "1                0.08            0.07            0.03            0.12   \n",
       "2               38.80           48.26           47.27           40.02   \n",
       "3                1.10            1.10            1.09            1.16   \n",
       "4                1.98            2.32            2.40            2.12   \n",
       "...               ...             ...             ...             ...   \n",
       "58802            0.29            0.28            0.36            0.19   \n",
       "58804            0.04            0.02            0.02            0.04   \n",
       "58830            0.77            0.68            0.62            0.62   \n",
       "58872            0.90            0.33            0.58            0.73   \n",
       "58879            0.34            0.22            0.21            0.56   \n",
       "\n",
       "       N2B27.D2_hNES1  N2B27.D3_hNES1  N2B27.D5_hNES1  PD.D1_hNES1  \\\n",
       "0                9.42            7.51            6.79         5.36   \n",
       "1                0.13            0.12            0.04         0.14   \n",
       "2               40.67           42.02           38.26        41.14   \n",
       "3                1.05            0.98            1.27         1.01   \n",
       "4                2.04            2.52            2.55         2.10   \n",
       "...               ...             ...             ...          ...   \n",
       "58802            0.28            0.31            0.55         0.22   \n",
       "58804            0.01            0.02            0.04         0.02   \n",
       "58830            0.59            0.56            0.77         0.58   \n",
       "58872            0.59            0.76            0.74         1.18   \n",
       "58879            0.21            0.31            0.30         0.32   \n",
       "\n",
       "       PD.D2_hNES1  PD.D3_hNES1  PD.D5_hNES1  PDA8301.D1_hNES1  \\\n",
       "0             6.30         6.97         5.86              6.00   \n",
       "1             0.05         0.07         0.18              0.02   \n",
       "2            40.60        37.82        38.43             43.27   \n",
       "3             1.15         1.56         1.96              1.11   \n",
       "4             2.31         3.36         3.73              2.17   \n",
       "...            ...          ...          ...               ...   \n",
       "58802         0.24         0.18         0.08              0.14   \n",
       "58804         0.02         0.03         0.02              0.02   \n",
       "58830         0.56         0.52         1.27              0.52   \n",
       "58872         0.89         1.10         1.14              1.12   \n",
       "58879         0.27         0.35         0.47              0.26   \n",
       "\n",
       "       PDA8301.D2_hNES1  PDA8301.D3_hNES1  PDA8301.D5_hNES1  \n",
       "0                  8.35              7.48              6.67  \n",
       "1                  0.04              0.12              0.08  \n",
       "2                 39.45             37.40             34.68  \n",
       "3                  1.19              1.82              1.86  \n",
       "4                  2.54              3.93              3.13  \n",
       "...                 ...               ...               ...  \n",
       "58802              0.09              0.07              0.05  \n",
       "58804              0.01              0.04              0.02  \n",
       "58830              0.41              0.64              1.70  \n",
       "58872              1.20              1.19              1.19  \n",
       "58879              0.28              0.53              0.46  \n",
       "\n",
       "[15454 rows x 19 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove all zero read data\n",
    "# FPKM data should be log2 - transformed for further analysis, and zero \n",
    "# values will cause errors\n",
    "# see the readme file if you would prefer to keep the zero read data in\n",
    "\n",
    "filtered_data = filtered_data[(filtered_data != 0).all(axis=1)]\n",
    "\n",
    "filtered_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f0bbc7e3-eee0-467d-abdc-768d322efdad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a list of the numeric columns \n",
    "# this can then be iterated through to apply the log2 function to the read data\n",
    "# there are fewer columns containing strings, so a list of string column titles is \n",
    "# made and subtracted from the total columns list\n",
    "\n",
    "col_titles = filtered_data.columns.tolist()\n",
    "remove_titles = [\"ensemblID\", \"biotype\", \"geneID\"] \n",
    "\n",
    "col_titles = [title for title in col_titles if title not in remove_titles]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d98ec3f0-ea62-41c6-beba-8762b31259c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>PDLGX_hNES1</th>\n",
       "      <th>A8301.D1_hNES1</th>\n",
       "      <th>A8301.D2_hNES1</th>\n",
       "      <th>A8301.D3_hNES1</th>\n",
       "      <th>N2B27.D1_hNES1</th>\n",
       "      <th>N2B27.D2_hNES1</th>\n",
       "      <th>N2B27.D3_hNES1</th>\n",
       "      <th>N2B27.D5_hNES1</th>\n",
       "      <th>PD.D1_hNES1</th>\n",
       "      <th>PD.D2_hNES1</th>\n",
       "      <th>PD.D3_hNES1</th>\n",
       "      <th>PD.D5_hNES1</th>\n",
       "      <th>PDA8301.D1_hNES1</th>\n",
       "      <th>PDA8301.D2_hNES1</th>\n",
       "      <th>PDA8301.D3_hNES1</th>\n",
       "      <th>PDA8301.D5_hNES1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>2.336283</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.107688</td>\n",
       "      <td>2.757023</td>\n",
       "      <td>3.035624</td>\n",
       "      <td>3.235727</td>\n",
       "      <td>2.908813</td>\n",
       "      <td>2.763412</td>\n",
       "      <td>2.422233</td>\n",
       "      <td>2.655352</td>\n",
       "      <td>2.801159</td>\n",
       "      <td>2.550901</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>3.061776</td>\n",
       "      <td>2.903038</td>\n",
       "      <td>2.737687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-3.643856</td>\n",
       "      <td>-3.836501</td>\n",
       "      <td>-5.058894</td>\n",
       "      <td>-3.058894</td>\n",
       "      <td>-2.943416</td>\n",
       "      <td>-3.058894</td>\n",
       "      <td>-4.643856</td>\n",
       "      <td>-2.836501</td>\n",
       "      <td>-4.321928</td>\n",
       "      <td>-3.836501</td>\n",
       "      <td>-2.473931</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-4.643856</td>\n",
       "      <td>-3.058894</td>\n",
       "      <td>-3.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>5.575010</td>\n",
       "      <td>5.277985</td>\n",
       "      <td>5.592756</td>\n",
       "      <td>5.562853</td>\n",
       "      <td>5.322649</td>\n",
       "      <td>5.345893</td>\n",
       "      <td>5.393004</td>\n",
       "      <td>5.257765</td>\n",
       "      <td>5.362470</td>\n",
       "      <td>5.343408</td>\n",
       "      <td>5.241077</td>\n",
       "      <td>5.264161</td>\n",
       "      <td>5.435295</td>\n",
       "      <td>5.301953</td>\n",
       "      <td>5.224966</td>\n",
       "      <td>5.116032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>0.176323</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.124328</td>\n",
       "      <td>0.214125</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>-0.029146</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.014355</td>\n",
       "      <td>0.201634</td>\n",
       "      <td>0.641546</td>\n",
       "      <td>0.970854</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.250962</td>\n",
       "      <td>0.863938</td>\n",
       "      <td>0.895303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>0.903038</td>\n",
       "      <td>0.985500</td>\n",
       "      <td>1.214125</td>\n",
       "      <td>1.263034</td>\n",
       "      <td>1.084064</td>\n",
       "      <td>1.028569</td>\n",
       "      <td>1.333424</td>\n",
       "      <td>1.350497</td>\n",
       "      <td>1.070389</td>\n",
       "      <td>1.207893</td>\n",
       "      <td>1.748461</td>\n",
       "      <td>1.899176</td>\n",
       "      <td>1.117695</td>\n",
       "      <td>1.344828</td>\n",
       "      <td>1.974529</td>\n",
       "      <td>1.646163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58802</th>\n",
       "      <td>AC055839.2</td>\n",
       "      <td>-3.184425</td>\n",
       "      <td>-1.785875</td>\n",
       "      <td>-1.836501</td>\n",
       "      <td>-1.473931</td>\n",
       "      <td>-2.395929</td>\n",
       "      <td>-1.836501</td>\n",
       "      <td>-1.689660</td>\n",
       "      <td>-0.862496</td>\n",
       "      <td>-2.184425</td>\n",
       "      <td>-2.058894</td>\n",
       "      <td>-2.473931</td>\n",
       "      <td>-3.643856</td>\n",
       "      <td>-2.836501</td>\n",
       "      <td>-3.473931</td>\n",
       "      <td>-3.836501</td>\n",
       "      <td>-4.321928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58804</th>\n",
       "      <td>AC069288.1</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-4.643856</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-4.643856</td>\n",
       "      <td>-6.643856</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-4.643856</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-5.058894</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-5.643856</td>\n",
       "      <td>-6.643856</td>\n",
       "      <td>-4.643856</td>\n",
       "      <td>-5.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58830</th>\n",
       "      <td>AC242843.1</td>\n",
       "      <td>-0.377070</td>\n",
       "      <td>-0.377070</td>\n",
       "      <td>-0.556393</td>\n",
       "      <td>-0.689660</td>\n",
       "      <td>-0.689660</td>\n",
       "      <td>-0.761213</td>\n",
       "      <td>-0.836501</td>\n",
       "      <td>-0.377070</td>\n",
       "      <td>-0.785875</td>\n",
       "      <td>-0.836501</td>\n",
       "      <td>-0.943416</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>-0.943416</td>\n",
       "      <td>-1.286304</td>\n",
       "      <td>-0.643856</td>\n",
       "      <td>0.765535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58872</th>\n",
       "      <td>AC022137.3</td>\n",
       "      <td>-0.217591</td>\n",
       "      <td>-0.152003</td>\n",
       "      <td>-1.599462</td>\n",
       "      <td>-0.785875</td>\n",
       "      <td>-0.454032</td>\n",
       "      <td>-0.761213</td>\n",
       "      <td>-0.395929</td>\n",
       "      <td>-0.434403</td>\n",
       "      <td>0.238787</td>\n",
       "      <td>-0.168123</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.189034</td>\n",
       "      <td>0.163499</td>\n",
       "      <td>0.263034</td>\n",
       "      <td>0.250962</td>\n",
       "      <td>0.250962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58879</th>\n",
       "      <td>AF196969.1</td>\n",
       "      <td>-1.643856</td>\n",
       "      <td>-1.556393</td>\n",
       "      <td>-2.184425</td>\n",
       "      <td>-2.251539</td>\n",
       "      <td>-0.836501</td>\n",
       "      <td>-2.251539</td>\n",
       "      <td>-1.689660</td>\n",
       "      <td>-1.736966</td>\n",
       "      <td>-1.643856</td>\n",
       "      <td>-1.888969</td>\n",
       "      <td>-1.514573</td>\n",
       "      <td>-1.089267</td>\n",
       "      <td>-1.943416</td>\n",
       "      <td>-1.836501</td>\n",
       "      <td>-0.915936</td>\n",
       "      <td>-1.120294</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15454 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           geneID  PDLGX_hNES1  A8301.D1_hNES1  A8301.D2_hNES1  \\\n",
       "0          TSPAN6     2.336283        3.000000        3.107688   \n",
       "1            TNMD    -5.643856       -3.643856       -3.836501   \n",
       "2            DPM1     5.575010        5.277985        5.592756   \n",
       "3           SCYL3     0.176323        0.137504        0.137504   \n",
       "4        C1orf112     0.903038        0.985500        1.214125   \n",
       "...           ...          ...             ...             ...   \n",
       "58802  AC055839.2    -3.184425       -1.785875       -1.836501   \n",
       "58804  AC069288.1    -5.643856       -4.643856       -5.643856   \n",
       "58830  AC242843.1    -0.377070       -0.377070       -0.556393   \n",
       "58872  AC022137.3    -0.217591       -0.152003       -1.599462   \n",
       "58879  AF196969.1    -1.643856       -1.556393       -2.184425   \n",
       "\n",
       "       A8301.D3_hNES1  N2B27.D1_hNES1  N2B27.D2_hNES1  N2B27.D3_hNES1  \\\n",
       "0            2.757023        3.035624        3.235727        2.908813   \n",
       "1           -5.058894       -3.058894       -2.943416       -3.058894   \n",
       "2            5.562853        5.322649        5.345893        5.393004   \n",
       "3            0.124328        0.214125        0.070389       -0.029146   \n",
       "4            1.263034        1.084064        1.028569        1.333424   \n",
       "...               ...             ...             ...             ...   \n",
       "58802       -1.473931       -2.395929       -1.836501       -1.689660   \n",
       "58804       -5.643856       -4.643856       -6.643856       -5.643856   \n",
       "58830       -0.689660       -0.689660       -0.761213       -0.836501   \n",
       "58872       -0.785875       -0.454032       -0.761213       -0.395929   \n",
       "58879       -2.251539       -0.836501       -2.251539       -1.689660   \n",
       "\n",
       "       N2B27.D5_hNES1  PD.D1_hNES1  PD.D2_hNES1  PD.D3_hNES1  PD.D5_hNES1  \\\n",
       "0            2.763412     2.422233     2.655352     2.801159     2.550901   \n",
       "1           -4.643856    -2.836501    -4.321928    -3.836501    -2.473931   \n",
       "2            5.257765     5.362470     5.343408     5.241077     5.264161   \n",
       "3            0.344828     0.014355     0.201634     0.641546     0.970854   \n",
       "4            1.350497     1.070389     1.207893     1.748461     1.899176   \n",
       "...               ...          ...          ...          ...          ...   \n",
       "58802       -0.862496    -2.184425    -2.058894    -2.473931    -3.643856   \n",
       "58804       -4.643856    -5.643856    -5.643856    -5.058894    -5.643856   \n",
       "58830       -0.377070    -0.785875    -0.836501    -0.943416     0.344828   \n",
       "58872       -0.434403     0.238787    -0.168123     0.137504     0.189034   \n",
       "58879       -1.736966    -1.643856    -1.888969    -1.514573    -1.089267   \n",
       "\n",
       "       PDA8301.D1_hNES1  PDA8301.D2_hNES1  PDA8301.D3_hNES1  PDA8301.D5_hNES1  \n",
       "0              2.584963          3.061776          2.903038          2.737687  \n",
       "1             -5.643856         -4.643856         -3.058894         -3.643856  \n",
       "2              5.435295          5.301953          5.224966          5.116032  \n",
       "3              0.150560          0.250962          0.863938          0.895303  \n",
       "4              1.117695          1.344828          1.974529          1.646163  \n",
       "...                 ...               ...               ...               ...  \n",
       "58802         -2.836501         -3.473931         -3.836501         -4.321928  \n",
       "58804         -5.643856         -6.643856         -4.643856         -5.643856  \n",
       "58830         -0.943416         -1.286304         -0.643856          0.765535  \n",
       "58872          0.163499          0.263034          0.250962          0.250962  \n",
       "58879         -1.943416         -1.836501         -0.915936         -1.120294  \n",
       "\n",
       "[15454 rows x 17 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a new dataframe with the identifier column preserved \n",
    "\n",
    "# ensemblID can be dropped, as it is an artefact of the data collection method\n",
    "# and does not store any additional information\n",
    "\n",
    "log2_data = pd.DataFrame()\n",
    "log2_data = filtered_data.iloc[:, 2:3]\n",
    "\n",
    "\n",
    "# add in the numerical values with the log2 function applied\n",
    "for col in col_titles: \n",
    "    log2_data[col] = np.log2(filtered_data[col])\n",
    "\n",
    "log2_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0192ef9d-e02b-41a7-9f56-720f72d96d94",
   "metadata": {},
   "source": [
    "# Data Exploration\n",
    "\n",
    "Now that we have a 'clean' dataset to work with, we can start to interrogate the data.\n",
    "\n",
    "First, we can run PCA to identify different lineage trajectories.\n",
    "\n",
    "Next, we can select some genes of interest for different cell lineages and examine the changes in their gene expression under different conditions.\n",
    "\n",
    "| Trophectoderm | Hypoblast | Core Epiblast | Primed |\n",
    "| --- | --- | --- | --- |\n",
    "| GATA2 | PDGFRA | NANOG | HLA-A |\n",
    "| GATA3 | FOXA2 | POU5F1 | DUSP6 | \n",
    "| NR2F2 | SOX17 | DPPA4| SOX11 |\n",
    "| TEAD1 | GATA4 | SOX2||\n",
    "| TEAD3 | COL4A1 | GDF3||\n",
    "| CLDN4 | COL4A2| DNMT3B||\n",
    "| OVOL1 ||||\n",
    "| KRT8 ||||\n",
    "| KRT18 ||||\n",
    "| HAND1 ||||\n",
    "| TP63 ||||\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79447c2-3944-4ac6-a458-a12b737a633f",
   "metadata": {},
   "source": [
    "# PCA Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1a995fa5-71bd-43fe-a69e-4758fec10109",
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare the total dataset for PCA\n",
    "scaler = StandardScaler()\n",
    "all_gene_ids = log2_data['geneID']\n",
    "\n",
    "all_log2_data = log2_data.drop(columns=['geneID'])\n",
    "\n",
    "all_data_scaled = scaler.fit_transform(all_log2_data)\n",
    "\n",
    "plot_data = pd.DataFrame()\n",
    "plot_data = all_data_scaled.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c8f6ba3e-3387-47ce-a968-b443d9b7caa2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Explained variance ratio: [0.53989128 0.20888743 0.06272463 0.04458007]\n",
      "Cumulative explained variance: [0.53989128 0.74877871 0.81150334 0.8560834 ]\n",
      "[[-0.00041523  0.00657009 -0.00143184 ...  0.00561327  0.00722444\n",
      "   0.00633165]\n",
      " [ 0.00118724  0.00426803 -0.00025386 ...  0.00519996 -0.00574389\n",
      "  -0.00137679]\n",
      " [ 0.0026622  -0.0108439  -0.00086596 ... -0.01154422  0.00117272\n",
      "  -0.00431863]\n",
      " [-0.00643445  0.0001995  -0.00179901 ... -0.00563189  0.0042738\n",
      "  -0.00338103]]\n"
     ]
    }
   ],
   "source": [
    "# run the PCA\n",
    "\n",
    "pca = PCA(n_components=4)\n",
    "pca_result = pca.fit_transform(plot_data)\n",
    "\n",
    "print(\"Explained variance ratio:\", pca.explained_variance_ratio_)\n",
    "print(\"Cumulative explained variance:\", np.cumsum(pca.explained_variance_ratio_))\n",
    "\n",
    "print(pca.components_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baf70707-174b-46bd-a52e-3bbaf6cdf5c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a plot to show which principle components explain the most variance in the data\n",
    "plt.bar(\n",
    "    range(1,len(pca.explained_variance_)+1),\n",
    "    (pca.explained_variance_ratio_)*100\n",
    "    )\n",
    " \n",
    " \n",
    "plt.xlabel('PCA Feature')\n",
    "plt.ylabel('Explained variance (%)')\n",
    "plt.title('Feature Explained Variance')\n",
    "plt.show()\n",
    "\n",
    "# this bar plot shows that Feature 1 and Feature 2 explain most of the variance, \n",
    "# with Features 3 and 4 contributing relatively little"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c67d41a8-920a-4715-b1fd-970a7a0fb9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function to map condition names to group (A8301, N2B27, PD03, PD03+A8301, PDLGX)\n",
    "def extract_condition_group(condition_name):\n",
    "    if 'PDLGX' in condition_name:\n",
    "        return 'PXGL' \n",
    "    elif 'PDA8301' in condition_name:\n",
    "        return 'PD03 and A83-01'\n",
    "    elif 'A8301' in condition_name:\n",
    "        return 'A83-01'\n",
    "    elif 'N2B27' in condition_name:\n",
    "        return 'N2B27'\n",
    "    elif 'PD' in condition_name: # PD must be lower in the if-tree than PDLGX or PDLGX conditions are falsely returned as PD \n",
    "        return 'PD03'\n",
    "    else:\n",
    "        return 'Other'  # For any other condition\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6de7e20-c868-4029-a763-1a3bb265c6fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to extract the time label (D1, D2, etc.) from condition names\n",
    "\n",
    "def extract_time_label(condition_name):\n",
    "    match = re.search(r'\\.D(\\d)', condition_name)  # Find the pattern \".D\" followed by a number to find the day number\n",
    "    if match:\n",
    "        return f'D{match.group(1)}'\n",
    "    elif 'PDLGX' in condition_name:\n",
    "        return \"D0\"\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1664034c-67aa-4a49-b2b4-e80e8ae08104",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract condition names from the columns of log2_data\n",
    "condition_names = all_log2_data.columns.tolist()  # Extract column names as the condition names\n",
    "\n",
    "# Map condition names to their corresponding groups\n",
    "condition_groups = [extract_condition_group(condition) for condition in condition_names]\n",
    "\n",
    "\n",
    "# Create a color map based on condition groups\n",
    "# Colours have been chosen to ensure that they are colourblind-friendly\n",
    "condition_colours = {\n",
    "    'PXGL': '#FF6347', \n",
    "    'N2B27': '#FFD700',  \n",
    "    'A83-01': '#9ACD32',  \n",
    "    'PD03': '#DA70D6',  \n",
    "    'PD03 and A83-01': '#4682B4',  \n",
    "    \n",
    "}\n",
    "\n",
    "colours = [condition_colours[group] for group in condition_groups]\n",
    "\n",
    "# Plot the PCA results, coloring the points based on their group\n",
    "plt.figure(figsize=(8, 6))\n",
    "\n",
    "\n",
    "scatter = plt.scatter(pca_result[:, 0], pca_result[:, 1], c=colours, edgecolor='black', alpha=0.7, s=100)\n",
    "\n",
    "# Add the figure legends\n",
    "plt.xlabel('First Principal Component')\n",
    "plt.ylabel('Second Principal Component')\n",
    "plt.grid(True)\n",
    "handles = []\n",
    "for group, color in condition_colours.items():\n",
    "    handles.append(plt.Line2D([0], [0], marker='o', color='w', label=group, markerfacecolor=color, markersize=10))\n",
    "\n",
    "plt.legend(handles=handles)\n",
    "\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a27cf765-7ea3-41d3-8a02-7b6be0bd8db7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### optional cell\n",
    "# this cell was made to produce a panel with both the PCA feature explanation\n",
    "# and the PCA itself for the report\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# Create a figure with 1 row and 2 columns for the panel\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 6))  # Adjust figsize as needed\n",
    "\n",
    "# --- Plot A: Explained Variance ---\n",
    "axes[0].bar(\n",
    "    range(1, len(pca.explained_variance_) + 1),\n",
    "    (pca.explained_variance_ratio_) * 100\n",
    ")\n",
    "\n",
    "axes[0].set_xlabel('PCA Feature')\n",
    "axes[0].set_ylabel('Explained variance (%)')\n",
    "\n",
    "# Add label \"A\" in the top-left corner of the first plot (A)\n",
    "axes[0].text(-0.1, 1.05, 'A', transform=axes[0].transAxes, fontsize=16, fontweight='bold')\n",
    "\n",
    "# --- Plot B: PCA Scatter Plot ---\n",
    "scatter = axes[1].scatter(pca_result[:, 0], pca_result[:, 1], c=colours, edgecolor='black', alpha=0.7, s=100)\n",
    "\n",
    "axes[1].set_xlabel('First Principal Component')\n",
    "axes[1].set_ylabel('Second Principal Component')\n",
    "axes[1].grid(True)\n",
    "\n",
    "# Add the figure legends on the second plot (B)\n",
    "handles = []\n",
    "for group, color in condition_colours.items():\n",
    "    handles.append(plt.Line2D([0], [0], marker='o', color='w', label=group, markerfacecolor=color, markersize=10))\n",
    "\n",
    "axes[1].legend(handles=handles)\n",
    "\n",
    "# Add label \"B\" in the top-left corner of the second plot (B)\n",
    "axes[1].text(-0.1, 1.05, 'B', transform=axes[1].transAxes, fontsize=16, fontweight='bold')\n",
    "\n",
    "# Adjust layout to ensure no overlap\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b50321af-243c-43ac-b34f-d739ea7845fc",
   "metadata": {},
   "source": [
    "# Graphing the different genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6613337c-4e90-423b-bfe1-8b8c8967bb55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the different marker subsets\n",
    "\n",
    "# trophectoderm subset\n",
    "TE_gene_list = [\"GATA2\", \"GATA3\", \"NR2F2\", \"TEAD1\", \"TEAD3\", \"CLDN4\", \"OVOL1\", \"KRT8\", \"KRT18\", \"HAND1\", \"TP63\"]\n",
    "\n",
    "# hypoblast subset\n",
    "HYP_gene_list = [\"PDGFRA\", \"FOXA2\", \"SOX17\", \"GATA4\", \"COL4A1\", \"COL4A2\"]\n",
    "\n",
    "# epiblast subset\n",
    "EPI_gene_list = [\"NANOG\", \"POU5F1\", \"DPPA4\", \"SOX2\", \"GDF3\", \"DNMT3B\", \"KLF4\", \"KLF17\", \"PRDM14\", \"TFAP2C\"]\n",
    "\n",
    "# primed subset\n",
    "PRI_gene_list = [\"HLA-C\", \"DUSP6\", \"SOX11\"]\n",
    "\n",
    "# all genes of interest\n",
    "GOI_gene_list = TE_gene_list + HYP_gene_list + EPI_gene_list + PRI_gene_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6201f8e7-af59-4b1e-a9f8-d0f2d7a85d52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to iterate through a list of strings\n",
    "# and pull out rows that contain that string in a given column\n",
    "\n",
    "def gene_subset(gene_list, data):\n",
    "    rows_to_add = []  # List to store rows to add\n",
    "    for gene in gene_list:\n",
    "        # Filter rows corresponding to the gene and append to the list\n",
    "        rows_to_add.append(data[data[\"geneID\"] == gene])\n",
    "\n",
    "    # Concatenate all dataframes from the list\n",
    "    subset_df = pd.concat(rows_to_add, ignore_index=True) if rows_to_add else pd.DataFrame()\n",
    "\n",
    "    return subset_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2497c3db-ca0d-45b5-b3bb-70dc05c7d5c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# use the defined function to pull out each of the genes of interest\n",
    "\n",
    "GOI_subset = pd.DataFrame()\n",
    "GOI_subset = gene_subset(GOI_gene_list, log2_data)\n",
    "GOI_subset.set_index(\"geneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c25b543d-4cdb-4af8-a292-b1e021988edb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Define the time points (shared for all genes)\n",
    "time_points = ['D0', 'D1', 'D2', 'D3', 'D5']\n",
    "GOI_subset.rename(columns={'PDLGX_hNES1': 'PDLGX.D0_hNES1'}, inplace=True)\n",
    "\n",
    "# Define the number of rows and columns for the grid of plots\n",
    "n_genes = len(GOI_subset.index)\n",
    "n_cols = 4  # Number of columns for the grid\n",
    "n_rows = (n_genes // n_cols) + (1 if n_genes % n_cols != 0 else 0)  # Calculate the number of rows needed\n",
    "                                                                    # for the list of genes\n",
    "\n",
    "# Create a figure with subplots\n",
    "fig, axes = plt.subplots(n_rows, n_cols, figsize=(15, 5 * n_rows))  # Adjust figure size depending on number \n",
    "                                                                    # of subplots in the grid\n",
    "\n",
    "axes = axes.flatten()\n",
    "\n",
    "\n",
    "# Loop through each gene to plot in the panels\n",
    "for i, gene in enumerate(GOI_subset.index):\n",
    "    ax = axes[i]  # Get the corresponding subplot for the current gene\n",
    "\n",
    "    condition_colors_used = {}  # Store the color for each condition group to avoid re-plotting the same group\n",
    "    \n",
    "    # Get PXGL Day 0 expression value for the current gene\n",
    "    pxgl_d0_expression = None\n",
    "    for condition in GOI_subset.columns:\n",
    "        if \"PDLGX\" in condition and \"D0\" in condition:  # Find PXGL D0 for the current gene\n",
    "            column_name = f'{condition.split(\".\")[0]}.D0_hNES1'\n",
    "            if column_name in GOI_subset.columns:\n",
    "                pxgl_d0_expression = GOI_subset.loc[gene, column_name]\n",
    "                break\n",
    "\n",
    "    # If PXGL D0 data was not found, skip this gene\n",
    "    if pxgl_d0_expression is None:\n",
    "        print(f\"Skipping gene {gene} because PXGL D0 expression is missing.\")\n",
    "        continue\n",
    "\n",
    "    # Loop through each condition and plot the data\n",
    "    for condition in GOI_subset.columns:\n",
    "      \n",
    "        # Extract the time point (e.g., D1, D2, D3, D5) from the condition name\n",
    "        time_label = extract_time_label(condition)\n",
    "\n",
    "        if time_label and time_label in time_points:\n",
    "            \n",
    "            # Extract the data for the current condition and gene for all time points\n",
    "            expression_values = []\n",
    "            for time_point in time_points:\n",
    "                \n",
    "                # Construct the expected column name\n",
    "                column_name = f'{condition.split(\".\")[0]}.{time_point}_hNES1'\n",
    "                \n",
    "                # Check if the column exists in the dataframe before trying to access it\n",
    "                if column_name in GOI_subset.columns:\n",
    "                    # Subtract PXGL D0 expression to make all time points relative to PXGL D0\n",
    "                    expression_values.append(GOI_subset.loc[gene, column_name] - pxgl_d0_expression)\n",
    "                else:\n",
    "                    expression_values.append(None)  # Append None if the column does not exist\n",
    "\n",
    "            # Only plot the condition group once, using the first time point to define the color\n",
    "            condition_group = extract_condition_group(condition)\n",
    "            \n",
    "            # Ensure we only plot the line once per condition group (first time)\n",
    "            if condition_group not in condition_colors_used:\n",
    "                # Get the color from the predefined condition_colours dictionary\n",
    "                color = condition_colours.get(condition_group, 'gray')  # Default to gray if no color is defined\n",
    "                condition_colors_used[condition_group] = color\n",
    "                \n",
    "                # Plot the line for the current condition\n",
    "                \n",
    "                ax.plot(time_points, expression_values, marker='o', label=condition_group, color=color, markersize=7)\n",
    "\n",
    "            # Connect D1 points to PXGL D0 for all conditions\n",
    "            if \"D1\" in time_points and pxgl_d0_expression is not None:\n",
    "                condition_d1_expression = expression_values[time_points.index('D1')]\n",
    "                if condition_d1_expression is not None:\n",
    "                    ax.plot([time_points.index('D0'), time_points.index('D1')], [0, condition_d1_expression], color=color, linestyle='--')\n",
    "   \n",
    "    # Add titles and legends\n",
    "    ax.set_title(f'Gene Expression for {GOI_subset.loc[gene, \"geneID\"]}')  \n",
    "    ax.set_xlabel('Time Point')\n",
    "    ax.set_ylabel('log2 Expression (relative to PXGL D0)')\n",
    "    ax.set_xticks(time_points)\n",
    "    ax.set_xticklabels(time_points, rotation=45)\n",
    "    ax.legend(title='Culture Conditions')\n",
    "    ax.grid(True)\n",
    "    ax.text(-0.1, 1.05, f'{i + 1}', transform=ax.transAxes, fontsize=12, fontweight='bold')\n",
    "    \n",
    "    \n",
    "# Hide any unused axes (if the number of genes is less than n_rows * n_cols)\n",
    "for x in range(n_genes, len(axes)):\n",
    "    axes[x].axis('off')\n",
    "\n",
    "# Ensure the plot layout fits nicely\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the plots\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23a61fe0-23b0-4916-a109-16cd60b22661",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare GOI_subset for plotting\n",
    "GOI_subset_plot = GOI_subset.drop(columns=['geneID'])\n",
    "GOI_subset_plot.index = GOI_subset['geneID'] \n",
    "\n",
    "\n",
    "\n",
    "# Create the heatmap \n",
    "plt.figure(figsize=(6, 8))  \n",
    "sns.heatmap(GOI_subset_plot, annot=False, cmap='coolwarm', cbar_kws={'label': 'Scaled log2 Expression'})\n",
    "\n",
    "# Set the title and labels\n",
    "plt.title('Gene Expression Heatmap for Gene of Interest Subset')\n",
    "\n",
    "plt.ylabel('Genes')  # Genes on the y-axis\n",
    "\n",
    "# Change the x axis labels so that they are meaningful to the viewer\n",
    "condition_labels = [f\"{extract_condition_group(col)} {extract_time_label(col)}\" for col in GOI_subset.columns]\n",
    "condition_labels[0] = ''\n",
    "condition_labels[1] = 'PXGL'\n",
    "\n",
    "# Rotate the tick labels for conditions (x-axis) and genes (y-axis)\n",
    "plt.xticks(ticks=range(len(GOI_subset.columns)), labels=condition_labels, rotation=45, ha='right') \n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
